internal/handler/middleware/middleware.go

```diff
@@ -27,8 +27,12 @@ import (
 
 // MsgRequest represents incoming request from the client.
 // It may come from the wire protocol connection or from the Data API server.
+//
+// It contains top level decoded document or an error from decoding OP_MSG request.
 type MsgRequest struct {
 	*wire.OpMsg
+	document *wirebson.Document
+	err      error
```

Comment:
If we can't decode the document, we should fail early, not propagate the error down

Reply:
The observability is handled downstream, so it's propagated down. However, failing decoding is rather unusual case.